File: python/examples/ex09_error_handling.py
"""Example 09 -- The error-handling matrix: every typed failure, deliberately.

Purpose
    Trigger each documented error class on purpose and show exactly what the
    shared client raises (typed exception, ``.code``, ``.request_id``,
    context extras) and what it RETRIED vs surfaced immediately:

    - 401 INVALID_API_KEY        -> AuthenticationError      (never retried)
    - 403 INSUFFICIENT_SCOPE     -> InsufficientScopeError   (never retried)
    - 402 INSUFFICIENT_CREDITS   -> InsufficientCreditsError (never retried;
      prints required vs balance and a top-up suggestion)
    - 422 VALIDATION_ERROR       -> ValidationError          (never retried)
    - 404 RUN_NOT_FOUND          -> NotFoundError            (never retried)
    - 409 NOT_AWAITING_HUMAN     -> ConflictError            (never retried)
    - 429 RATE_LIMITED           -> client honors Retry-After, then recovers
    - 503/500 server errors      -> retried w/ backoff, then ServerError
    - 500 on a POST WITHOUT an Idempotency-Key -> NOT retried (safety guard)

Flow
    Each scenario is a small function taking ``(base_url, api_key)``; it
    builds its own client with a recording (non-sleeping) ``sleep`` and a
    request-counting hook, performs a doomed (or recovering) call, and
    returns a :class:`ScenarioResult` row. ``run_all`` renders the matrix.
    Tests point ``base_url`` at respx mocks; against the real production API
    this example only PRINTS the matrix (which scenarios are live-safe) and
    never fires requests.

    Some failures (403 scope, 402 wallet, forced 5xx) cannot be triggered
    organically on a healthy server. Against the cookbook's local mock
    server (``make mock``) those scenarios send the documented
    ``X-Mock-Force-Error: <CODE>`` test hook; ``main`` probes for that
    support with one free request and skips the forceable scenarios when
    the server does not honor it. The 429-then-recover demo needs a
    scripted 429->200 sequence, so it always defers to the test suite.

Endpoints
    GET /v1/models, POST /v1/predict, POST /v1/runs, GET /v1/runs/{id},
    POST /v1/runs/{id}/resume, POST /v1/machines/{id}/terminal

Estimated cost
    $0.00 in practice -- every scenario is designed to FAIL (and predict
    charges are auto-refunded on failure; see coasty.cost / the pricing
    table). No spend gate needed: nothing here can complete a billable op
    except the 429-recovery demo, which only runs against injected mocks.
"""

from __future__ import annotations

import argparse
import random
import sys
import uuid
from collections.abc import Callable, Sequence
from dataclasses import dataclass

import httpx

from coasty import (
    CoastyClient,
    CoastyError,
    InsufficientCreditsError,
    InsufficientScopeError,
    env,
)

# Obviously-fake keys: scenarios never touch the real API.
DEMO_API_KEY = "sk-coasty-test-" + "0" * 48
# Malformed prefix: rejected as 401 by ANY conforming server (the documented
# key families are sk-coasty-live-/sk-coasty-test-/cua_sk_ only).
BAD_API_KEY = "sk-coasty-bogus-" + "f" * 48
FAKE_SCREENSHOT_B64 = "iVBORw0KGgoAAAANSUhEUg" + "A" * 120  # >100 chars, no data: prefix
FORCE_ERROR_HEADER = "X-Mock-Force-Error"  # cookbook mock-server test hook


@dataclass(frozen=True)
class ScenarioResult:
    """One row of the matrix: what happened when we poked the failure."""

    name: str
    outcome: str  # "raised" | "recovered"
    exception: str | None
    code: str | None
    request_id: str | None
    status_code: int | None
    attempts: int
    slept: tuple[float, ...]
    detail: str

    @property
    def client_retried(self) -> bool:
        # The client always backs off (sleeps) between retry attempts, so
        # recorded sleeps -- not request count -- distinguish a retry from a
        # scenario that legitimately makes several different requests.
        return len(self.slept) > 0


ScenarioFn = Callable[[str, str, "dict[str, str] | None"], ScenarioResult]


@dataclass(frozen=True)
class Scenario:
    """A named, self-contained failure demonstration."""

    name: str
    description: str
    safe_live: bool  # free + read-only: OK to fire at the real API
    run: ScenarioFn
    # Error code to force via the mock server's X-Mock-Force-Error hook when
    # the failure cannot be triggered organically (None = organic scenario).
    force_error: str | None = None
    # True when only a scripted response sequence (e.g. 429 then 200) can
    # demonstrate the behavior -- always deferred to the test suite.
    needs_script: bool = False


def _instrumented_client(
    base_url: str, api_key: str, extra_headers: dict[str, str] | None = None
) -> tuple[CoastyClient, httpx.Client, list[float], list[str]]:
    """A client whose sleeps are recorded (never real) and requests counted."""
    sleeps: list[float] = []
    requests_seen: list[str] = []

    def _count(request: httpx.Request) -> None:
        requests_seen.append(f"{request.method} {request.url.path}")

    http_client = httpx.Client(event_hooks={"request": [_count]}, headers=extra_headers or {})
    client = CoastyClient(
        api_key=api_key,
        base_url=base_url,
        http_client=http_client,
        sleep=sleeps.append,  # deterministic: record, never block
        rng=random.Random(7),  # deterministic backoff jitter
    )
    return client, http_client, sleeps, requests_seen


def _run_one(
    name: str,
    base_url: str,
    api_key: str,
    call: Callable[[CoastyClient], str],
    *,
    detail_for: Callable[[CoastyError], str] | None = None,
    extra_headers: dict[str, str] | None = None,
) -> ScenarioResult:
    """Run one doomed (or recovering) call and capture the evidence."""
    client, http_client, sleeps, requests_seen = _instrumented_client(
        base_url, api_key, extra_headers
    )
    try:
        recovered_detail = call(client)
    except CoastyError as exc:
        return ScenarioResult(
            name=name,
            outcome="raised",
            exception=type(exc).__name__,
            code=exc.code,
            request_id=exc.request_id,
            status_code=exc.status_code,
            attempts=len(requests_seen),
            slept=tuple(sleeps),
            detail=detail_for(exc) if detail_for is not None else exc.message,
        )
    finally:
        client.close()
        http_client.close()
    return ScenarioResult(
        name=name,
        outcome="recovered",
        exception=None,
        code=None,
        request_id=None,
        status_code=200,
        attempts=len(requests_seen),
        slept=tuple(sleeps),
        detail=recovered_detail,
    )


# ── scenario functions ─────────────────────────────────────────────────────


def scenario_invalid_api_key(
    base_url: str, api_key: str, extra_headers: dict[str, str] | None = None
) -> ScenarioResult:
    """401: a bad key on a free endpoint. AuthenticationError, no retry."""
    del api_key  # this scenario deliberately uses a (fake) bad key

    def call(client: CoastyClient) -> str:
        client.models()
        return "unexpected success"

    return _run_one("401 INVALID_API_KEY", base_url, BAD_API_KEY, call, extra_headers=extra_headers)


def scenario_insufficient_scope(
    base_url: str, api_key: str, extra_headers: dict[str, str] | None = None
) -> ScenarioResult:
    """403: key lacks terminal:exec. Shows required_scope vs current_scopes."""

    def call(client: CoastyClient) -> str:
        client.machine_terminal("mch_demo", "whoami")
        return "unexpected success"

    def detail(exc: CoastyError) -> str:
        if isinstance(exc, InsufficientScopeError):
            return (
                f"needs scope {exc.required_scope!r}, key has {exc.current_scopes!r} -- "
                "mint a key with the right scopes"
            )
        return exc.message

    return _run_one(
        "403 INSUFFICIENT_SCOPE",
        base_url,
        api_key,
        call,
        detail_for=detail,
        extra_headers=extra_headers,
    )


def scenario_insufficient_credits(
    base_url: str, api_key: str, extra_headers: dict[str, str] | None = None
) -> ScenarioResult:
    """402: wallet cannot cover the op. Prints required vs balance + top-up."""

    def call(client: CoastyClient) -> str:
        client.predict(FAKE_SCREENSHOT_B64, "open the settings page")
        return "unexpected success"

    def detail(exc: CoastyError) -> str:
        if isinstance(exc, InsufficientCreditsError):
            required = exc.required if exc.required is not None else 0
            balance = exc.balance if exc.balance is not None else 0
            shortfall = max(0, required - balance)
            return (
                f"required={required} cr, balance={balance} cr -- "
                f"top up at least ${shortfall / 100:.2f} to proceed"
            )
        return exc.message

    return _run_one(
        "402 INSUFFICIENT_CREDITS",
        base_url,
        api_key,
        call,
        detail_for=detail,
        extra_headers=extra_headers,
    )


def scenario_validation_error(
    base_url: str, api_key: str, extra_headers: dict[str, str] | None = None
) -> ScenarioResult:
    """422: an invalid body (empty instruction / short screenshot). No retry."""

    def call(client: CoastyClient) -> str:
        client.predict("too-short", "")
        return "unexpected success"

    return _run_one("422 VALIDATION_ERROR", base_url, api_key, call, extra_headers=extra_headers)


def scenario_not_found(
    base_url: str, api_key: str, extra_headers: dict[str, str] | None = None
) -> ScenarioResult:
    """404: ids are mode-isolated; a missing run raises NotFoundError."""

    def call(client: CoastyClient) -> str:
        client.get_run("run_does_not_exist")
        return "unexpected success"

    return _run_one("404 RUN_NOT_FOUND", base_url, api_key, call, extra_headers=extra_headers)


def scenario_not_awaiting_human(
    base_url: str, api_key: str, extra_headers: dict[str, str] | None = None
) -> ScenarioResult:
    """409: resume is only valid from awaiting_human. Shows current_state.

    Organic flow: create a run (which starts ``queued``/``running``) and
    immediately try to resume it -- the server must refuse with
    NOT_AWAITING_HUMAN because nothing is paused. Two requests, no retries.
    """

    def call(client: CoastyClient) -> str:
        created = client.create_run(
            "mch_demo",
            "reconcile the invoices (this run is never resumed from a pause)",
            idempotency_key=f"ex09-409-{uuid.uuid4().hex[:8]}",
        )
        client.resume_run(str(created.data["id"]))
        return "unexpected success"

    def detail(exc: CoastyError) -> str:
        current = exc.extras.get("current_state")
        allowed = exc.extras.get("allowed_from")
        return f"run is {current!r}; resume allowed only from {allowed!r}"

    return _run_one(
        "409 NOT_AWAITING_HUMAN",
        base_url,
        api_key,
        call,
        detail_for=detail,
        extra_headers=extra_headers,
    )


def scenario_rate_limited_recovers(
    base_url: str, api_key: str, extra_headers: dict[str, str] | None = None
) -> ScenarioResult:
    """429 then 200: the client sleeps exactly Retry-After, then succeeds.

    /predict is safe to retry (charged-then-refunded on failure), so the
    client transparently honors the server's pacing.
    """

    def call(client: CoastyClient) -> str:
        result = client.predict(FAKE_SCREENSHOT_B64, "open the settings page")
        return (
            f"recovered after rate limit (request_id={result.request_id}, "
            f"credits_charged={result.credits_charged})"
        )

    return _run_one(
        "429 RATE_LIMITED (recovers)", base_url, api_key, call, extra_headers=extra_headers
    )


def scenario_server_errors_surface(
    base_url: str, api_key: str, extra_headers: dict[str, str] | None = None
) -> ScenarioResult:
    """Persistent 5xx: retried with backoff (Retry-After honored), then raised.

    Failed predictions are charged-then-auto-refunded, so the surfaced
    ServerError costs nothing.
    """

    def call(client: CoastyClient) -> str:
        client.predict(FAKE_SCREENSHOT_B64, "open the settings page")
        return "unexpected success"

    return _run_one(
        "503/500 retry-then-surface", base_url, api_key, call, extra_headers=extra_headers
    )


def scenario_unsafe_post_not_retried(
    base_url: str, api_key: str, extra_headers: dict[str, str] | None = None
) -> ScenarioResult:
    """500 on POST /runs WITHOUT an Idempotency-Key: surfaced on attempt 1.

    The client only retries POSTs that are inherently safe (predict/ground/
    parse) or carry an Idempotency-Key -- otherwise a retry could start the
    same run twice.
    """

    def call(client: CoastyClient) -> str:
        client.create_run("mch_demo", "reconcile the invoices")  # no idempotency_key
        return "unexpected success"

    return _run_one(
        "500 unsafe POST (no retry)", base_url, api_key, call, extra_headers=extra_headers
    )


SCENARIOS: tuple[Scenario, ...] = (
    Scenario(
        "401 INVALID_API_KEY",
        "Bad key on GET /models -> AuthenticationError; 4xx are never retried.",
        safe_live=True,
        run=scenario_invalid_api_key,
    ),
    Scenario(
        "403 INSUFFICIENT_SCOPE",
        "terminal:exec missing -> InsufficientScopeError with required_scope.",
        safe_live=False,
        run=scenario_insufficient_scope,
        force_error="INSUFFICIENT_SCOPE",
    ),
    Scenario(
        "402 INSUFFICIENT_CREDITS",
        "Wallet too small for /predict -> required vs balance + top-up hint.",
        safe_live=False,
        run=scenario_insufficient_credits,
        force_error="INSUFFICIENT_CREDITS",
    ),
    Scenario(
        "422 VALIDATION_ERROR",
        "Empty instruction + short screenshot -> ValidationError with details.",
        safe_live=True,
        run=scenario_validation_error,
    ),
    Scenario(
        "404 RUN_NOT_FOUND",
        "GET a run id that does not exist -> NotFoundError.",
        safe_live=True,
        run=scenario_not_found,
    ),
    Scenario(
        "409 NOT_AWAITING_HUMAN",
        "Resume a run that is not paused -> ConflictError with current_state.",
        safe_live=False,
        run=scenario_not_awaiting_human,
    ),
    Scenario(
        "429 RATE_LIMITED (recovers)",
        "429 with Retry-After then 200 -> the client sleeps exactly that long.",
        safe_live=False,
        run=scenario_rate_limited_recovers,
        needs_script=True,
    ),
    Scenario(
        "503/500 retry-then-surface",
        "Persistent 5xx -> backoff retries (max 4 attempts) then ServerError.",
        safe_live=False,
        run=scenario_server_errors_surface,
        force_error="UPSTREAM_UNAVAILABLE",
    ),
    Scenario(
        "500 unsafe POST (no retry)",
        "POST /runs without Idempotency-Key -> surfaced on the first attempt.",
        safe_live=False,
        run=scenario_unsafe_post_not_retried,
        force_error="INTERNAL_ERROR",
    ),
)


def _skipped(scenario: Scenario, why: str) -> ScenarioResult:
    return ScenarioResult(
        name=scenario.name,
        outcome="skipped",
        exception=None,
        code=None,
        request_id=None,
        status_code=None,
        attempts=0,
        slept=(),
        detail=why,
    )


def run_all(
    base_url: str, api_key: str = DEMO_API_KEY, *, mock_hooks: bool = False
) -> list[ScenarioResult]:
    """Run every scenario against ``base_url`` (a mock or local mock server).

    ``mock_hooks=True`` means the server honors the cookbook mock's
    ``X-Mock-Force-Error`` header, which lets us trigger the failures that a
    healthy server never produces organically (403/402/forced 5xx).
    """
    rows: list[ScenarioResult] = []
    for scenario in SCENARIOS:
        if scenario.needs_script:
            rows.append(
                _skipped(scenario, "needs a scripted 429->200 sequence; see the respx tests")
            )
        elif scenario.force_error is not None and not mock_hooks:
            rows.append(
                _skipped(
                    scenario,
                    f"cannot trigger organically; needs {FORCE_ERROR_HEADER} support "
                    "(run the cookbook mock server) or the respx tests",
                )
            )
        elif scenario.force_error is not None:
            rows.append(scenario.run(base_url, api_key, {FORCE_ERROR_HEADER: scenario.force_error}))
        else:
            rows.append(scenario.run(base_url, api_key, None))
    return rows


def format_results(results: Sequence[ScenarioResult]) -> str:
    """Render the matrix: exception, code, request_id, attempts, sleeps."""
    lines = [
        f"{'scenario':<28} {'outcome':<10} {'exception':<26} {'code':<22} "
        f"{'attempts':>8} {'request_id':<14} detail"
    ]
    for row in results:
        retry_note = f"{row.attempts}{'*' if row.client_retried else ''}"
        lines.append(
            f"{row.name:<28} {row.outcome:<10} {row.exception or '-':<26} "
            f"{row.code or '-':<22} {retry_note:>8} {row.request_id or '-':<14} {row.detail}"
        )
    lines.append("(* = the client retried; sleeps were recorded, never real)")
    return "\n".join(lines)


def format_catalog() -> str:
    """The static matrix + live-safety notes (printed in live mode)."""
    lines = ["Error-handling matrix (live mode: scenarios are NOT fired at production):"]
    for scenario in SCENARIOS:
        safety = "live-safe (free, read-only)" if scenario.safe_live else "mock-only"
        lines.append(f"  - {scenario.name:<28} [{safety}] {scenario.description}")
    lines.append("Point COASTY_BASE_URL at the local mock server (make mock) to execute them all.")
    return "\n".join(lines)


# ── CLI ────────────────────────────────────────────────────────────────────


def _supports_mock_hooks(base_url: str, api_key: str) -> bool:
    """One free probe: does this server honor the X-Mock-Force-Error hook?

    GET /models is free; a server that honors the hook must answer with the
    forced NOT_FOUND envelope instead of the model list.
    """
    try:
        response = httpx.get(
            f"{base_url}/models",
            headers={"X-API-Key": api_key, FORCE_ERROR_HEADER: "NOT_FOUND"},
            timeout=10.0,
        )
        body = response.json()
    except (httpx.HTTPError, ValueError):
        return False
    error = body.get("error") if isinstance(body, dict) else None
    return (
        response.status_code == 404 and isinstance(error, dict) and error.get("code") == "NOT_FOUND"
    )


def main(argv: Sequence[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__.split("\n", maxsplit=1)[0])
    parser.add_argument(
        "--base-url",
        default=None,
        help="override the target (default: COASTY_BASE_URL or the production URL)",
    )
    args = parser.parse_args(argv)

    base_url = (args.base_url or env.get_base_url()).rstrip("/")
    if base_url == env.DEFAULT_BASE_URL:
        # Never fire deliberate failures at production: just print the matrix.
        print(format_catalog())
        return 0

    api_key = env.get_api_key() or DEMO_API_KEY
    mock_hooks = _supports_mock_hooks(base_url, api_key)
    print(
        f"{FORCE_ERROR_HEADER} hook: "
        + ("supported -- executing forceable scenarios too" if mock_hooks else "not supported")
    )
    try:
        results = run_all(base_url, api_key, mock_hooks=mock_hooks)
    except CoastyError as exc:  # a scenario failed in an UNexpected way
        print(
            f"unexpected API error {exc.code} (request_id={exc.request_id}): {exc.message}",
            file=sys.stderr,
        )
        return 1
    print(format_results(results))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: typescript/src/examples/ex09-error-handling.ts
/**
 * Example 09 — Error-handling matrix.
 *
 * One scenario per row of the documented error catalog, each demonstrating
 * the typed error the shared client raises and the context extras it carries:
 *
 *   | scenario              | HTTP | code                  | typed class              |
 *   | --------------------- | ---- | --------------------- | ------------------------ |
 *   | invalid_api_key       | 401  | INVALID_API_KEY       | AuthenticationError      |
 *   | insufficient_scope    | 403  | INSUFFICIENT_SCOPE    | InsufficientScopeError   |
 *   | insufficient_credits  | 402  | INSUFFICIENT_CREDITS  | InsufficientCreditsError |
 *   | validation_error      | 422  | VALIDATION_ERROR      | ValidationError          |
 *   | not_found             | 404  | RUN_NOT_FOUND         | NotFoundError            |
 *   | not_awaiting_human    | 409  | NOT_AWAITING_HUMAN    | ConflictError            |
 *   | rate_limited          | 429  | RATE_LIMITED          | RateLimitError           |
 *   | internal_error        | 500  | INTERNAL_ERROR        | ServerError              |
 *   | upstream_unavailable  | 503  | UPSTREAM_UNAVAILABLE  | ServerError              |
 *
 * Branch on `error.code` (stable), never on `message`. 429/500/503/504 are
 * retried by the client (exponential backoff + full jitter, base 0.5s, cap
 * 8s, max 4 attempts, `Retry-After` honored) and only then surfaced; other
 * 4xx are NEVER retried. Every error carries `request_id` — quote it to
 * support.
 *
 * Estimated cost: 0 credits ($0.00). Every scenario either fails before
 * billing or hits an inference failure that is charged-then-auto-refunded.
 * Safety: pointed at the PRODUCTION base URL this example only LISTS the
 * scenarios; it executes them only against an explicitly overridden
 * `COASTY_BASE_URL` (e.g. the offline mock server at http://127.0.0.1:8787/v1).
 *
 * Run it:
 *   COASTY_BASE_URL=http://127.0.0.1:8787/v1 npx tsx src/examples/ex09-error-handling.ts
 */
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { CoastyClient, type CoastyClientOptions } from '../coasty/client.js';
import { DEFAULT_BASE_URL, getApiKey, getBaseUrl } from '../coasty/env.js';
import { CoastyError } from '../coasty/errors.js';

// ---------------------------------------------------------------------------
// Scenario catalog
// ---------------------------------------------------------------------------

/** > 100 chars of base64-looking padding, as /predict requires. */
const DEMO_SCREENSHOT = 'iVBORw0KGgo'.padEnd(120, 'A');

/** An obviously fake, syntactically invalid key for the 401 demo. */
export const OBVIOUSLY_INVALID_KEY = 'sk-coasty-test-invalid-key-for-the-401-demo';

export interface ErrorScenario {
  name: string;
  description: string;
  expectedStatus: number;
  expectedCode: string;
  /** The class name of the typed error the client raises. */
  expectedClass: string;
  /** True when the client retries before surfacing this error. */
  retried: boolean;
  /** Per-scenario client tweaks (e.g. the broken key for the 401 demo). */
  clientOptions?: Partial<CoastyClientOptions>;
  /** Fire the request that produces the documented failure. */
  trigger: (client: CoastyClient) => Promise<unknown>;
}

export const SCENARIOS: readonly ErrorScenario[] = [
  {
    name: 'invalid_api_key',
    description: 'Any call with a malformed/unknown key',
    expectedStatus: 401,
    expectedCode: 'INVALID_API_KEY',
    expectedClass: 'AuthenticationError',
    retried: false,
    clientOptions: { apiKey: OBVIOUSLY_INVALID_KEY },
    trigger: (client) => client.models(),
  },
  {
    name: 'insufficient_scope',
    description: 'GET /machines/{id}/connection without connection:read',
    expectedStatus: 403,
    expectedCode: 'INSUFFICIENT_SCOPE',
    expectedClass: 'InsufficientScopeError',
    retried: false,
    trigger: (client) => client.machines.connection('mch_test_demo'),
  },
  {
    name: 'insufficient_credits',
    description: 'POST /runs with a wallet below one step',
    expectedStatus: 402,
    expectedCode: 'INSUFFICIENT_CREDITS',
    expectedClass: 'InsufficientCreditsError',
    retried: false,
    trigger: (client) =>
      client.runs.create({ machine_id: 'mch_test_demo', task: 'Reconcile the invoices' }),
  },
  {
    name: 'validation_error',
    description: 'POST /predict with an empty instruction',
    expectedStatus: 422,
    expectedCode: 'VALIDATION_ERROR',
    expectedClass: 'ValidationError',
    retried: false,
    trigger: (client) => client.predict({ screenshot: DEMO_SCREENSHOT, instruction: '' }),
  },
  {
    name: 'not_found',
    description: 'GET /runs/{id} for a run that does not exist',
    expectedStatus: 404,
    expectedCode: 'RUN_NOT_FOUND',
    expectedClass: 'NotFoundError',
    retried: false,
    trigger: (client) => client.runs.get('run_does_not_exist'),
  },
  {
    name: 'not_awaiting_human',
    description: 'POST /runs/{id}/resume on a run that is not paused',
    expectedStatus: 409,
    expectedCode: 'NOT_AWAITING_HUMAN',
    expectedClass: 'ConflictError',
    retried: false,
    trigger: (client) => client.runs.resume('run_not_paused', { note: 'resume attempt' }),
  },
  {
    name: 'rate_limited',
    description: 'POST /predict over the rate limit (Retry-After honored, then surfaced)',
    expectedStatus: 429,
    expectedCode: 'RATE_LIMITED',
    expectedClass: 'RateLimitError',
    retried: true,
    trigger: (client) => client.predict({ screenshot: DEMO_SCREENSHOT, instruction: 'go' }),
  },
  {
    name: 'internal_error',
    description: 'POST /predict hitting a 500 (retried, auto-refunded, then surfaced)',
    expectedStatus: 500,
    expectedCode: 'INTERNAL_ERROR',
    expectedClass: 'ServerError',
    retried: true,
    trigger: (client) => client.predict({ screenshot: DEMO_SCREENSHOT, instruction: 'go' }),
  },
  {
    name: 'upstream_unavailable',
    description: 'POST /ground during an upstream outage (503 + Retry-After)',
    expectedStatus: 503,
    expectedCode: 'UPSTREAM_UNAVAILABLE',
    expectedClass: 'ServerError',
    retried: true,
    trigger: (client) => client.ground({ screenshot: DEMO_SCREENSHOT, element: 'Submit button' }),
  },
];

// ---------------------------------------------------------------------------
// Normalization + runner (testable core)
// ---------------------------------------------------------------------------

export interface ScenarioReport {
  scenario: string;
  /** True when the raised error matched the documented code AND class. */
  matched: boolean;
  errorClass: string;
  code: string;
  status: number | null;
  /** Always surfaced — quote it to support. */
  requestId: string | null;
  /** Human-readable context extras pulled off the typed error. */
  notes: string[];
}

/** Pull the documented context extras off a typed error into note lines. */
export function describeError(error: CoastyError): string[] {
  const notes: string[] = [];
  if (error.requiredScope !== undefined) {
    notes.push(
      `required scope: ${error.requiredScope}` +
        (error.currentScopes === undefined ? '' : ` (current: ${error.currentScopes.join(', ')})`),
    );
  }
  if (error.required !== undefined || error.balance !== undefined) {
    notes.push(
      `required ${String(error.required ?? '?')} cr vs balance ${String(error.balance ?? '?')} cr`,
    );
  }
  if (error.retryAfter !== undefined) {
    notes.push(`retry_after ${String(error.retryAfter)}s — the client slept exactly that long`);
  }
  if (error.currentState !== undefined) {
    notes.push(
      `current state: ${error.currentState}` +
        (error.allowedFrom === undefined ? '' : ` (allowed from: ${error.allowedFrom.join(', ')})`),
    );
  }
  if (error.details !== undefined) notes.push(`details: ${JSON.stringify(error.details)}`);
  if (error.suggestion !== undefined) notes.push(`suggestion: ${error.suggestion}`);
  return notes;
}

/** Run one scenario and normalize whatever it raised. Never throws on CoastyError. */
export async function runScenario(
  scenario: ErrorScenario,
  client: CoastyClient,
): Promise<ScenarioReport> {
  try {
    await scenario.trigger(client);
    return {
      scenario: scenario.name,
      matched: false,
      errorClass: '(none)',
      code: '(none)',
      status: null,
      requestId: null,
      notes: ['expected an error but the call succeeded'],
    };
  } catch (error) {
    if (!(error instanceof CoastyError)) throw error;
    const notes = describeError(error);
    if (scenario.retried) {
      notes.push('retried with exponential backoff + full jitter before surfacing');
    }
    return {
      scenario: scenario.name,
      matched:
        error.code === scenario.expectedCode && error.constructor.name === scenario.expectedClass,
      errorClass: error.constructor.name,
      code: error.code,
      status: error.statusCode,
      requestId: error.requestId,
      notes,
    };
  }
}

export interface RunAllOptions {
  /** Injectable target — tests point this at a mocked transport. */
  baseUrl: string;
  apiKey?: string;
  /** Forwarded to every per-scenario client (fetch/sleep/random injection). */
  clientOptions?: Partial<CoastyClientOptions>;
  logger?: (line: string) => void;
}

/** Build the per-scenario client (scenario overrides win over shared options). */
export function buildScenarioClient(scenario: ErrorScenario, options: RunAllOptions): CoastyClient {
  return new CoastyClient({
    baseUrl: options.baseUrl,
    ...(options.apiKey === undefined ? {} : { apiKey: options.apiKey }),
    ...options.clientOptions,
    ...scenario.clientOptions,
  });
}

export async function runAllScenarios(options: RunAllOptions): Promise<ScenarioReport[]> {
  const log = options.logger ?? ((line: string): void => void process.stdout.write(`${line}\n`));
  const reports: ScenarioReport[] = [];
  for (const scenario of SCENARIOS) {
    const report = await runScenario(scenario, buildScenarioClient(scenario, options));
    reports.push(report);
    log(formatReportLine(report));
    for (const note of report.notes) log(`      ${note}`);
  }
  return reports;
}

export function formatReportLine(report: ScenarioReport): string {
  const mark = report.matched ? 'OK ' : 'MISMATCH';
  return (
    `${mark} ${report.scenario.padEnd(22)} ${String(report.status ?? '-').padStart(3)} ` +
    `${report.code.padEnd(22)} ${report.errorClass.padEnd(26)} request_id=${report.requestId ?? 'n/a'}`
  );
}

// ---------------------------------------------------------------------------
// Thin CLI — live mode only LISTS; execution needs a non-production base URL
// ---------------------------------------------------------------------------

function isMain(): boolean {
  const entry = process.argv[1];
  return entry !== undefined && path.resolve(entry) === fileURLToPath(import.meta.url);
}

export async function main(): Promise<void> {
  const print = (line: string): void => void process.stdout.write(`${line}\n`);
  const baseUrl = getBaseUrl();

  if (baseUrl === DEFAULT_BASE_URL) {
    print('Pointed at the production API — listing the scenario catalog only (no calls made).');
    print('Set COASTY_BASE_URL to the offline mock server to execute them, e.g.:');
    print('  COASTY_BASE_URL=http://127.0.0.1:8787/v1  (make mock)');
    print('');
    for (const scenario of SCENARIOS) {
      print(
        `  ${scenario.name.padEnd(22)} ${String(scenario.expectedStatus)} ` +
          `${scenario.expectedCode.padEnd(22)} -> ${scenario.expectedClass.padEnd(26)} ${scenario.description}`,
      );
    }
    return;
  }

  print(`executing ${String(SCENARIOS.length)} scenarios against ${baseUrl} (cost: $0.00)`);
  const reports = await runAllScenarios({ baseUrl, apiKey: getApiKey() });
  const mismatched = reports.filter((report) => !report.matched);
  if (mismatched.length > 0) {
    console.error(`${String(mismatched.length)} scenario(s) did not match the documented catalog`);
    process.exitCode = 1;
  }
}

if (isMain()) {
  main().catch((error: unknown) => {
    if (error instanceof CoastyError) {
      console.error(`[${error.code}] ${error.message} (request_id: ${error.requestId ?? 'n/a'})`);
    } else {
      console.error(String(error));
    }
    process.exitCode = 1;
  });
}
